umbraco/Umbraco-CMS
# Support segments from querystring in preview

### Prerequisites


### Description

When previewing documents, one can appy a `culture` parameter to the querystring to change the culture being previewed. This is what ultimately makes the language selector work in preview mode.

For some odd reason, segments have never been supported in the same way, which is what this PR fixes.

Note that only the "by ID" and "by key" content finders support this kind of context switching by means of the querystring ... the default content finder (for resolving via route) does not.

### Testing this PR

1. Configure a few languages.
2. Setup a page using a template that prints out the current variation context (see example below). It does not need to be a variant page for this to work.
3. Preview the page.
4. The preview frame (UI) does not yet support segments, so you'll have to exit the preview frame by clicking this button:
![image](https://github.com/user-attachments/assets/21008f98-f28a-4c2d-81c6-e5b6d00a93e2)
5. Now verify that you can change culture _and_ segment by using the `culture` and `segment` querystring parameters, respectively.

#### Template example

```cshtml
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage
@inject IVariationContextAccessor VariationContextAccessor;
@{
    Layout = null;
    var context = VariationContextAccessor.VariationContext ?? throw new InvalidOperationException("No variant context found!");
}
<html lang="en">
<body>
<h1>@Model.Name</h1>
<ul>
    <li>Culture: @context.Culture</li>
    <li>Segment: @context.Segment</li>
</ul>
</body>
</html>
```
